lib/kube/proxy/testing/kube_server/kube_mock.go

```diff
@@ -527,3 +538,82 @@ func (s *KubeMockServer) selfSubjectAccessReviews(w http.ResponseWriter, req *ht
 
 	return s1, nil
 }
+
+// portforward supports SPDY protocols only. Teleport always uses SPDY when
+// portforwarding to upstreams even if the original request is WebSocket.
+func (s *KubeMockServer) portforward(w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	_, err = httpstream.Handshake(req, w, []string{PortForwardProtocolV1Name})
+	if err != nil {
+		return
+	}
+
+	streamChan := make(chan httpstream.Stream, 1)
+
+	upgrader := spdystream.NewResponseUpgraderWithPings(defaults.HighResPollingPeriod)
+	conn := upgrader.UpgradeResponse(w, req, httpStreamReceived(req.Context(), streamChan))
+	if conn == nil {
+		err = trace.ConnectionProblem(nil, "unable to upgrade SPDY connection")
+		return
+	}
+	defer conn.Close()
+	var (
+		data      httpstream.Stream
+		errStream httpstream.Stream
+	)
+
+	for {
+		select {
+		case <-conn.CloseChan():
+			return
+		case stream := <-streamChan:
+			streamType := stream.Headers().Get(StreamType)
+			switch streamType {
+			case StreamTypeError:
+				errStream = stream
+			case StreamTypeData:
+				data = stream
+			}
+		}
+		if errStream != nil && data != nil {
+			break
+		}
+	}
+
+	buf := make([]byte, 1024)
+	n, err := data.Read(buf)
+	if err != nil {
+		errStream.Write([]byte(err.Error()))
+	}
+	data.Write([]byte(fmt.Sprintf(PortForwardPayload, p.ByName("podName"), string(buf[:n]))))
+	return
+}
+
+// httpStreamReceived is the httpstream.NewStreamHandler for port
+// forward streams. It checks each stream's port and stream type headers,
+// rejecting any streams that with missing or invalid values. Each valid
+// stream is sent to the streams channel.
+func httpStreamReceived(ctx context.Context, streams chan httpstream.Stream) func(httpstream.Stream, <-chan struct{}) error {
+	return func(stream httpstream.Stream, replySent <-chan struct{}) error {
+		// make sure it has a valid port header
+		portString := stream.Headers().Get(PortHeader)
+		if len(portString) == 0 {
+			return trace.BadParameter("%q header is required", PortHeader)
+		}
+
+		// make sure it has a valid stream type header
+		streamType := stream.Headers().Get(StreamType)
+		if len(streamType) == 0 {
+			return trace.BadParameter("%q header is required", StreamType)
+		}
+		if streamType != StreamTypeError && streamType != StreamTypeData {
+			return trace.BadParameter("invalid stream type %q", streamType)
+		}
+
+		select {
+		case streams <- stream:
+			return nil
+		case <-ctx.Done():
+			return trace.BadParameter("request has been canceled")
```

Comment:
Typically we'd pass ctx.Err() forward:

```suggestion
			return trace.Wrap(ctx.Err())
```